xde/src/xde.rs

```diff
@@ -1320,50 +1465,107 @@ fn clear_xde_underlay() -> Result<NoResp, OpteError> {
     // Clear multicast forwarding table to release any references
     token.mcast_fwd.write().clear();
 
-    if let Some(underlay) = token.underlay.take() {
-        // If the underlay references have leaked/spread beyond `XdeDev`s and not
-        // been cleaned up, we committed have a fatal programming error.
-        // We aren't using `Weak` references to these types either, so no strong
-        // references could be created.
-        //
-        // We know these must succeed given that the only holders of an
-        // `Arc<XdeUnderlayPort>` are `XdeState` (whose ref we have exclusively locked)
-        // and `XdeDev` (of which none remain).
-        let name = underlay.u1.name.clone();
-        let u1 = Arc::into_inner(underlay.u1).unwrap_or_else(|| {
-            panic!("underlay u1 ({name}) must have one ref during teardown",)
-        });
+    // Before taking ownership of the underlay Arcs, clear per-CPU cached
+    // `DevMap`s and multicast forwarding tables (in underlay ports' `PerEntryState`).
+    // This breaks snapshot cycles: underlay → per-CPU cache → `DevMap` →
+    // `XdeDev` → underlay.
+    //
+    // Note: Per-port `DevMap` caches (`XdeDev.port_map`) were already cleared
+    // when ports were deleted. This function only runs after all ports are
+    // removed.
+    if let Some(ul_ref) = token.underlay.as_ref() {
+        let empty_map = Arc::new(DevMap::new());
+        let empty_mcast: Arc<McastForwardingTable> = Arc::new(BTreeMap::new());
+        let underlay_ports =
+            [&ul_ref.u1.stream.ports_map, &ul_ref.u2.stream.ports_map];
+        for per_cpu_map in underlay_ports {
+            for entry in per_cpu_map {
+                {
+                    let mut map = entry.devs.lock();
+                    *map = Arc::clone(&empty_map);
+                }
+                {
+                    let mut mcast = entry.mcast_fwd.write();
+                    *mcast = Arc::clone(&empty_mcast);
+                }
+                entry.has_mcast_subscribers.store(false, Ordering::Relaxed);
+            }
+        }
+    }
```

Comment:
I don't see why we have this here. If `token.devs` is empty and we have the management lock, all `DevMaps` seen by `PerEntryState` should also be empty by definition -- any `delete`s have succeeded, and no one else can attempt a write to `token.devs` in parallel. So, there really should be no other references to the underlay. If they're not empty, that's a fatal programming error on our part?